### PR TITLE
fix: Fixed-Point `pow` Now Uses Binary Exponentiation

### LINKED ISSUE #272: [L-04]: Fixed-Point pow Uses Linear Exponentiation, Accumulating Unnecessary Rounding Error

### PR DESCRIPTION
**Pull request by 0xNeshi**
Switches `UD30x9` and `SD29x9` `pow` from linear repeated multiplication to binary exponentiation, addressing the performance and accumulated truncation concerns from issue #272.

**Changes:**
- Simplified `UD30x9::pow` and `SD29x9::pow` to delegate to the shared helper, eliminating per-iteration casts and redundant loop code.
- Updated `pow` documentation in both modules to clarify truncation and associativity implications under binary exponentiation grouping.
- Changed the `SD29x9` overflow test to use the minimal failing case `sd29x9::max().pow(3)` (fails the `base_mag <= MIN_NEGATIVE_VALUE` assertion).

#### PR Checklist

- [x] Tests
- [x] Documentation
- [ ] Changelog

## Summary by CodeRabbit

* **Refactor**
  * Updated power calculation implementation to use an optimized exponentiation algorithm with improved overflow handling. The computation of intermediate steps may differ from previous versions, particularly in edge cases with high exponents and fixed-point truncation.

* **Tests**
  * Refactored test assertions for power functions to verify updated computation behavior and edge case handling.

### CODE REVIEW

File: math/fixed_point/tests/sd29x9_tests/pow_tests.move [hunk 1]
```diff
@@ -47,16 +47,14 @@ fun pow_handles_negative_one_parity() {
...
-    expect(val.pow(255), val.pow(254).mul(val));
+    // 1.25^16 * 10^9 = 35_527_136_787 before flooring
+    expect(val.pow(16), pos(35_527_136_781));
+    expect(val.pow(255), pos(5_152_918_999_790_606_401_120_741_084_983_548));
 }
 
 #[test, expected_failure(abort_code = sd29x9_base::EOverflow)]
 fun pow_overflow_aborts_for_large_base() {
-    sd29x9::max().pow(2);
+    sd29x9::max().pow(32);
```
Comment (0xNeshi) [inline math/fixed_point/tests/sd29x9_tests/pow_tests.move]: Covers the [limit assertion code](https://github.com/OpenZeppelin/contracts-sui/pull/281/changes#diff-95359e7450872a0bf549130c17a487e208e005465a962f94548425d74b6f7c36R364)

File: math/fixed_point/sources/sd29x9/sd29x9_base.move [hunk 2]
```diff
@@ -320,27 +320,28 @@ public fun div(x: SD29x9, y: SD29x9): SD29x9 {
...
+///
 /// #### Parameters
 /// - `x`: Base value.
 /// - `exp`: Exponent.
 ///
 /// #### Returns
-/// - An approximation of `x^exp` using the same stepwise truncation semantics as repeated
-///   fixed-point multiplication.
+/// - An approximation of `x^exp` computed using binary exponentiation and fixed-point truncation.
 ///
```
[the remark this answers is not in the document]
Comment (0xNeshi) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move:342]: This hasn't been released yet

File: math/fixed_point/sources/ud30x9/ud30x9_base.move [hunk 3]
```diff
@@ -292,39 +292,49 @@ public fun div(x: UD30x9, y: UD30x9): UD30x9 {
...
 ///
 /// As a consequence, `pow` is approximate for most fractional values: rounding error compounds as
 /// `exp` grows, results are biased toward zero, and for `0 < x < 1` intermediate values can reach
 /// zero before the final mathematically scaled result would.
 ///
+/// Because truncation is applied at intermediate steps, the result generally matches neither the
+/// exact real-valued power rounded once at the end nor the result of left-to-right repeated
+/// multiplication. In particular, fixed-point multiplication is not associative under truncation,
+/// so the grouping of operations used by binary exponentiation affects the final value.
+///
```
[the remark this answers is not in the document]
Comment (0xNeshi) [reply math/fixed_point/sources/ud30x9/ud30x9_base.move:306]: Hasn't been released yet

File: math/fixed_point/tests/ud30x9_tests/pow_tests.move [hunk 4]
```diff
@@ -38,7 +38,9 @@ fun pow_handles_fractional_values_and_truncation() {
 #[test]
 fun pow_supports_high_exponents() {
     let val = fixed(SCALE + 250_000_000); // 1.25
-    val.pow(255);
+    expect(val.pow(255), pos(5_152_918_999_790_606_401_120_741_084_983_548));
+    // with binary exponentiation, rounding/truncation behavior for larger exponents is affected by grouping
+    expect_ne!(val.pow(255), val.pow(254).mul(val));
```
Comment (0xNeshi) [inline math/fixed_point/tests/ud30x9_tests/pow_tests.move]:
```suggestion
    let pow255 = val.pow(255);
    expect(pow255, pos(5_152_918_999_790_606_401_120_741_084_983_548));
    // with binary exponentiation, rounding/truncation behavior for larger exponents is affected by grouping
    expect_ne!(pow255, val.pow(254).mul(val));
```

File: math/fixed_point/tests/ud30x9_tests/pow_tests.move [hunk 5]
```diff
@@ -38,7 +38,10 @@ fun pow_handles_fractional_values_and_truncation() {
 #[test]
 fun pow_supports_high_exponents() {
     let val = fixed(SCALE + 250_000_000); // 1.25
-    val.pow(255);
+    let pow255 = val.pow(255);
+    expect(pow255, pos(5_152_918_999_790_606_401_120_741_084_983_548));
```
Comment (0xNeshi) [inline math/fixed_point/tests/ud30x9_tests/pow_tests.move]:
```suggestion
    expect(pow255, fixed(5_152_918_999_790_606_401_120_741_084_983_548));
```

File: math/fixed_point/tests/ud30x9_tests/helpers.move
Comment (bidzyyys) [inline math/fixed_point/tests/ud30x9_tests/helpers.move:1]: I do not like `expect` nor `expect_ne` functions. Check my feedback from PR #282.

File: math/fixed_point/tests/sd29x9_tests/helpers.move
Comment (bidzyyys) [inline math/fixed_point/tests/sd29x9_tests/helpers.move:1]: Use `assert_eq` from Move unit tests when possible.

File: math/fixed_point/tests/sd29x9_tests/pow_tests.move [hunk 6]
```diff
@@ -47,16 +47,14 @@ fun pow_handles_negative_one_parity() {
...
-    expect(val.pow(255), val.pow(254).mul(val));
+    // Expected value is the result of sd29x9::pow's binary exponentiation with intermediate truncation, not simply floor(1.25^16 * 10^9) (the exact value before flooring is 35_527_136_787)
+    expect(val.pow(16), pos(35_527_136_781));
+    expect(val.pow(255), pos(5_152_918_999_790_606_401_120_741_084_983_548));
 }
 
 #[test, expected_failure(abort_code = sd29x9_base::EOverflow)]
 fun pow_overflow_aborts_for_large_base() {
-    sd29x9::max().pow(2);
+    sd29x9::max().pow(32);
```
[the remark this answers is not in the document]
Comment (bidzyyys) [reply math/fixed_point/tests/sd29x9_tests/pow_tests.move]: @0xNeshi looks like a fair point from @copilot.

File: math/fixed_point/tests/sd29x9_tests/pow_tests.move [hunk 7]
```diff
@@ -3,7 +3,7 @@ module openzeppelin_fp_math::sd29x9_pow_tests;
 
 use openzeppelin_fp_math::sd29x9;
 use openzeppelin_fp_math::sd29x9_base;
-use openzeppelin_fp_math::sd29x9_test_helpers::{pos, neg, expect};
+use openzeppelin_fp_math::sd29x9_test_helpers::{pos, neg, expect, expect_ne};
```
Comment (bidzyyys) [inline math/fixed_point/tests/sd29x9_tests/pow_tests.move]: Do not like `expect` nor `expect_ne`.

File: math/fixed_point/tests/sd29x9_tests/pow_tests.move [hunk 8]
```diff
@@ -47,16 +47,18 @@ fun pow_handles_negative_one_parity() {
 #[test]
 fun pow_supports_high_exponents() {
     let val = pos(SCALE + 250_000_000); // 1.25
-    // Exact result for exp=16: sequential truncation of (res_mag * 1_250_000_000 / 1_000_000_000)
-    // gives 35_527_136_770 (vs. exact 1.25^16 * 10^9 = 35_527_136_787 before flooring).
-    expect(val.pow(16), pos(35_527_136_770));
-    // Recurrence invariant for the maximum u8 exponent: pow(n) == pow(n-1).mul(base).
-    expect(val.pow(255), val.pow(254).mul(val));
+    // Expected value is the result of sd29x9::pow's binary exponentiation with intermediate truncation, not simply floor(1.25^16 * 10^9) (the exact value before flooring is 35_527_136_787)
```
Comment (bidzyyys) [inline math/fixed_point/tests/sd29x9_tests/pow_tests.move]: Shouldn't you use `sd29x9::pow` etc ? I mean `'`

File: math/fixed_point/sources/ud30x9/ud30x9_base.move [hunk 9]
```diff
@@ -292,39 +292,49 @@ public fun div(x: UD30x9, y: UD30x9): UD30x9 {
...
-    let mut result = base;
-    let times = exp - 1;
-    times.do!(|_| {
-        result = result * base / SCALE_U256;
-        assert!(result <= max_value, EOverflow);
-    });
+    let mut base = x.unwrap() as u256;
+    let mut result = SCALE_U256;
+
+    while (exp != 0) {
```
Comment (bidzyyys) [inline math/fixed_point/sources/ud30x9/ud30x9_base.move:329]: Isn't it the same code for `SD` & `UD` data types? If yes, can we keep it only in a single place/file? Would be much easier to maintain.

Comment (bidzyyys) [reply math/fixed_point/sources/ud30x9/ud30x9_base.move:329]: In terms of different error for different data type -> You can return `option::none` from helper and map it to `EOverflow` in you proper file.

File: math/fixed_point/sources/sd29x9/sd29x9_base.move [hunk 10]
```diff
@@ -349,12 +351,21 @@ public fun pow(x: SD29x9, exp: u8): SD29x9 {
...
+
+    while (exp != 0) {
+        if ((exp & 1) == 1) {
+            res_mag = res_mag * base_mag / SCALE;
+            assert!(res_mag <= MIN_NEGATIVE_VALUE as u256, EOverflow);
+        };
+        exp = exp >> 1;
+        if (exp != 0) {
+            base_mag = base_mag * base_mag / SCALE;
+            assert!(base_mag <= MIN_NEGATIVE_VALUE as u256, EOverflow);
```
Comment (bidzyyys) [inline math/fixed_point/sources/sd29x9/sd29x9_base.move:366]: Casting needed for every loop iteration. Can we optimise this?

File: math/fixed_point/sources/sd29x9/sd29x9_base.move [hunk 11]
```diff
@@ -349,12 +351,21 @@ public fun pow(x: SD29x9, exp: u8): SD29x9 {
...
-        res_mag = res_mag * mag / SCALE;
-        assert!(res_mag <= MIN_NEGATIVE_VALUE as u256, EOverflow);
-    });
+    let mut res_mag = SCALE;
+    let mut base_mag = mag;
+
+    while (exp != 0) {
+        if ((exp & 1) == 1) {
+            res_mag = res_mag * base_mag / SCALE;
+            assert!(res_mag <= MIN_NEGATIVE_VALUE as u256, EOverflow);
```
Comment (bidzyyys) [inline math/fixed_point/sources/sd29x9/sd29x9_base.move:361]: Casting needed for every loop iteration. Can we optimise this?

**Review by bidzyyys — COMMENTED** [review]
Left some comments.

File: math/fixed_point/tests/sd29x9_tests/pow_tests.move (hunk 8 above)
Comment (0xNeshi) [reply math/fixed_point/tests/sd29x9_tests/pow_tests.move]: @bidzyyys unclear what you meant?

File: math/fixed_point/sources/ud30x9/ud30x9_base.move (hunk 9 above)
Comment (0xNeshi) [reply math/fixed_point/sources/ud30x9/ud30x9_base.move:329]: This won't work once https://github.com/OpenZeppelin/contracts-sui/pull/280 is merged, because the base*base assertion will be slightly different. Copilot updated the code prematurely.

We'll revisit this once the other PR is merged

File: math/fixed_point/sources/sd29x9/sd29x9_base.move (hunk 10 above)
Comment (0xNeshi) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move:366]: This was already addressed in https://github.com/OpenZeppelin/contracts-sui/pull/280

File: math/fixed_point/sources/sd29x9/sd29x9_base.move (hunk 11 above)
Comment (0xNeshi) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move:361]: This was already addressed in https://github.com/OpenZeppelin/contracts-sui/pull/280

File: math/fixed_point/sources/sd29x9/sd29x9_base.move (hunk 10 above)
Comment (0xNeshi) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move:366]: Copilot again updated the code prematurely, this is annoying

File: math/fixed_point/tests/sd29x9_tests/pow_tests.move (hunk 6 above)
[the remark this answers is not in the document]
Comment (0xNeshi) [reply math/fixed_point/tests/sd29x9_tests/pow_tests.move]: @bidzyyys partially accurate suggestion. The minimum exponent is actually `3`, updating the code

File: math/fixed_point/sources/sd29x9/sd29x9_base.move (hunk 10 above)
Comment (ericnordelo) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move:366]: Why is copilot updating code at all?

File: math/fixed_point/tests/sd29x9_tests/pow_tests.move [hunk 12]
```diff
@@ -3,60 +3,64 @@ module openzeppelin_fp_math::sd29x9_pow_tests;
...
+    assert_eq!(neg_base.pow(3).unwrap(), neg(8 * SCALE).unwrap());
 }
 
 #[test]
 fun pow_handles_fractional_values_and_truncation() {
     // 1.5^2 = 2.25, 1.5^3 = 3.375
     let one_point_five = pos(1_500_000_000);
-    expect(one_point_five.pow(2), pos(2_250_000_000));
-    expect(one_point_five.pow(3), pos(3_375_000_000));
+    assert_eq!(one_point_five.pow(2).unwrap(), pos(2_250_000_000).unwrap());
```
Comment (ericnordelo) [inline math/fixed_point/tests/sd29x9_tests/pow_tests.move]: Should we keep expect as a macro?

Comment (0xNeshi) [reply math/fixed_point/tests/sd29x9_tests/pow_tests.move]: `.unwrap()` is redundant, removing it. So `expect` is redundant too (see group chat).

Copilot prematurely implemented this, creating more work...

Comment (ericnordelo) [reply math/fixed_point/tests/sd29x9_tests/pow_tests.move]: Got it.

Comment (ericnordelo) [reply math/fixed_point/tests/sd29x9_tests/pow_tests.move]: Again, not sure why is copilot updating anything. Did we change a config somewhere?

File: math/fixed_point/tests/sd29x9_tests/pow_tests.move [hunk 13]
```diff
@@ -3,60 +3,64 @@ module openzeppelin_fp_math::sd29x9_pow_tests;
...
+    assert_eq!(pow255.unwrap(), pos(5_152_918_999_790_606_401_120_741_084_983_548).unwrap());
+    // with binary exponentiation, rounding/truncation behavior for larger exponents is affected by grouping
+    assert!(pow255.unwrap() != val.pow(254).mul(val).unwrap());
 }
 
 #[test, expected_failure(abort_code = sd29x9_base::EOverflow)]
 fun pow_overflow_aborts_for_large_base() {
-    sd29x9::max().pow(2);
+    sd29x9::max().pow(3);
 }
```
[the remark this answers is not in the document]
Comment (0xNeshi) [reply math/fixed_point/tests/sd29x9_tests/pow_tests.move:64]: 3 is the minimum exponent that triggers the failure

File: math/fixed_point/sources/sd29x9/sd29x9_base.move (hunk 10 above)
Comment (0xNeshi) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move:366]: I'm wondering the same thing. I just lost 15 mins reverting its changes.

File: math/fixed_point/tests/sd29x9_tests/pow_tests.move (hunk 12 above)
Comment (0xNeshi) [reply math/fixed_point/tests/sd29x9_tests/pow_tests.move]: Not that I know. @bidzyyys ?

File: math/fixed_point/tests/ud30x9_tests/pow_tests.move [hunk 14]
```diff
@@ -4,6 +4,7 @@ module openzeppelin_fp_math::ud30x9_pow_tests;
 use openzeppelin_fp_math::ud30x9;
 use openzeppelin_fp_math::ud30x9_base;
 use openzeppelin_fp_math::ud30x9_test_helpers::{fixed, expect};
+use std::unit_test::assert_eq;
```
[the remark this answers is not in the document]
Comment (0xNeshi) [reply math/fixed_point/tests/ud30x9_tests/pow_tests.move:7]: Addressed in https://github.com/OpenZeppelin/contracts-sui/pull/282

File: math/fixed_point/tests/sd29x9_tests/pow_tests.move [hunk 15]
```diff
@@ -4,6 +4,7 @@ module openzeppelin_fp_math::sd29x9_pow_tests;
 use openzeppelin_fp_math::sd29x9;
 use openzeppelin_fp_math::sd29x9_base;
 use openzeppelin_fp_math::sd29x9_test_helpers::{pos, neg, expect};
+use std::unit_test::assert_eq;
```
[the remark this answers is not in the document]
Comment (0xNeshi) [reply math/fixed_point/tests/sd29x9_tests/pow_tests.move:7]: addressed in https://github.com/OpenZeppelin/contracts-sui/pull/282

File: math/fixed_point/sources/ud30x9/ud30x9_base.move [hunk 16]
```diff
@@ -292,39 +292,50 @@ public fun div(x: UD30x9, y: UD30x9): UD30x9 {
...
+        if ((exp & 1) == 1) {
+            result = result * base / SCALE_U256;
+            assert!(result <= max_value, EOverflow);
+        };
+        exp = exp >> 1;
+        if (exp != 0) {
+            base = base * base / SCALE_U256;
+            assert!(base <= max_value, EOverflow);
+        };
+    };
```
[the remark this answers is not in the document]
Comment (0xNeshi) [reply math/fixed_point/sources/ud30x9/ud30x9_base.move]: See https://github.com/OpenZeppelin/contracts-sui/pull/281#discussion_r3014961245

File: math/fixed_point/sources/sd29x9/sd29x9_base.move [hunk 17]
```diff
@@ -349,12 +351,21 @@ public fun pow(x: SD29x9, exp: u8): SD29x9 {
...
+        };
+        exp = exp >> 1;
+        if (exp != 0) {
+            base_mag = base_mag * base_mag / SCALE;
+            assert!(base_mag <= MIN_NEGATIVE_VALUE as u256, EOverflow);
+        };
+    };
```
[the remark this answers is not in the document]
Comment (0xNeshi) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move:368]: See https://github.com/OpenZeppelin/contracts-sui/pull/281#discussion_r3014961245

**Review by bidzyyys — APPROVED** [review]

File: math/fixed_point/sources/ud30x9/ud30x9_base.move [hunk 18]
```diff
@@ -292,39 +292,50 @@ public fun div(x: UD30x9, y: UD30x9): UD30x9 {
...
 ///
 /// #### Returns
-/// - An approximation of `x^exp` using the same stepwise truncation semantics as repeated
-///   fixed-point multiplication.
+/// - An approximation of `x^exp` computed using binary exponentiation and fixed-point truncation.
 ///
 /// #### Aborts
 /// - Aborts if the resulting value exceeds the representable `UD30x9` range.
-public fun pow(x: UD30x9, exp: u8): UD30x9 {
+public fun pow(x: UD30x9, mut exp: u8): UD30x9 {
```
Comment (ericnordelo) [inline math/fixed_point/sources/ud30x9/ud30x9_base.move]:
```suggestion
public fun pow(x: UD30x9, exp: u8): UD30x9 {
```

I don't think this function should mofidy the given exp, we can shadow it locally instead.

File: math/fixed_point/sources/sd29x9/sd29x9_base.move [hunk 19]
```diff
@@ -320,27 +320,29 @@ public fun div(x: SD29x9, y: SD29x9): SD29x9 {
...
 ///
 /// #### Returns
-/// - An approximation of `x^exp` using the same stepwise truncation semantics as repeated
-///   fixed-point multiplication.
+/// - An approximation of `x^exp` computed using binary exponentiation and fixed-point truncation.
 ///
 /// #### Aborts
 /// - Aborts if the resulting magnitude exceeds the representable `SD29x9` range.
-public fun pow(x: SD29x9, exp: u8): SD29x9 {
+public fun pow(x: SD29x9, mut exp: u8): SD29x9 {
```
Comment (bidzyyys) [inline math/fixed_point/sources/sd29x9/sd29x9_base.move]: Same as [here](https://github.com/OpenZeppelin/contracts-sui/pull/281/changes#r3016133135)

**Review by ericnordelo — APPROVED** [review]
LGTM!

It would be great to merge [this PR](https://github.com/OpenZeppelin/contracts-sui/pull/264/changes#diff-87b41c3729172495815119446202a3ed99d10702c45c30413a6a302efeb61710) first to avoid conflicts with constants, but non blocking by any means.
